components/calendar/src/japanese.rs

```diff
@@ -0,0 +1,219 @@
+    fn year(&self, date: &Self::DateInner) -> types::Year {
+        types::Year {
+            era: types::Era(date.era),
+            number: date.adjusted_year(),
+            related_iso: date.inner.year.0,
+        }
+    }
+
+    /// The calendar-specific month represented by `date`
+    fn month(&self, date: &Self::DateInner) -> types::Month {
+        Iso.month(&date.inner)
+    }
+
+    /// The calendar-specific day-of-month represented by `date`
+    fn day_of_month(&self, date: &Self::DateInner) -> types::DayOfMonth {
+        Iso.day_of_month(&date.inner)
+    }
+
+    /// Information of the day of the year
+    fn day_of_year_info(&self, date: &Self::DateInner) -> types::DayOfYearInfo {
+        let prev_dec_31 = IsoDateInner::dec_31((date.inner.year.0 - 1).into());
+        let next_jan_1 = IsoDateInner::jan_1((date.inner.year.0 + 1).into());
+
+        let prev_dec_31 = self.date_from_iso(Date::from_raw(prev_dec_31, Iso));
+        let next_jan_1 = self.date_from_iso(Date::from_raw(next_jan_1, Iso));
+        types::DayOfYearInfo {
+            day_of_year: Iso::day_of_year(date.inner),
+            days_in_year: Iso::days_in_year(date.inner.year),
+            prev_year: self.year(&prev_dec_31),
+            days_in_prev_year: Iso::days_in_year(prev_dec_31.inner.year),
+            next_year: self.year(&next_jan_1),
+        }
+    }
+
+    fn debug_name() -> &'static str {
+        "Japanese"
+    }
+}
+
+impl JapaneseDateInner {
+    /// Returns the current year relative to the era
+    fn adjusted_year(&self) -> i32 {
+        // The year in which an era starts is Year 1, and it may be short
+        // The only time this function will experience dates that are *before*
+        // the era start date are for the first era (Currently, taika-645),
+        // where we elect to still report the year as year 1 when it is in the same
+        // Gregorian year, and use zero/negative years before that.
+        self.inner.year.0 - self.era_start + 1
+    }
+}
+
+const MEIJI_START: EraStartDate = EraStartDate {
+    year: 1868,
+    month: 9,
+    day: 8,
+};
+const TAISHO_START: EraStartDate = EraStartDate {
+    year: 1912,
+    month: 7,
+    day: 30,
+};
+const SHOWA_START: EraStartDate = EraStartDate {
+    year: 1926,
+    month: 12,
+    day: 25,
+};
+const HEISEI_START: EraStartDate = EraStartDate {
+    year: 1989,
+    month: 1,
+    day: 8,
+};
+const REIWA_START: EraStartDate = EraStartDate {
+    year: 2019,
+    month: 5,
+    day: 1,
+};
+
+const FALLBACK_ERA: (EraStartDate, TinyStr16) = (REIWA_START, tinystr16!("reiwa"));
+
+impl Japanese {
+    /// Given an ISO date, obtain the era data
+    fn era_for(&self, date: &IsoDateInner) -> (EraStartDate, TinyStr16) {
+        let date: EraStartDate = date.into();
+        let era_data = self.eras.get();
+        let deref_tuple = |x: (&EraStartDate, &TinyStr16)| (*x.0, *x.1);
+        if date >= MEIJI_START {
+            if era_data.dates_to_eras.len() == 5 {
```

Comment:
`question`: how did you chose 5 as a cut-off for hardcoded eras? Can you document your thinking to help someone one day follow the logics when updating it if needed?

Reply:
Oh, "Meiji" onwards are the five modern eras.